include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BXDF_NDF_GGX_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BXDF_NDF_GGX_INCLUDED_
+
+#include "nbl/builtin/hlsl/limits.hlsl"
+#include "nbl/builtin/hlsl/bxdf/ndf.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace bxdf
+{
+namespace ndf
+{
+
+// TODO: use query_type when that's implemented
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T>)
+struct GGX<T,false NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+{
+    using scalar_type = T;
+    using this_t = GGX<T,false>;
+
+    // trowbridge-reitz
+    scalar_type D(scalar_type a2, scalar_type NdotH2)
+    {
+        scalar_type denom = NdotH2 * (a2 - scalar_type(1.0)) + scalar_type(1.0);
+        return a2 * numbers::inv_pi<scalar_type> / (denom * denom);
+    }
+
+    scalar_type FVNDF(scalar_type fresnel_ndf, scalar_type G1_over_2NdotV, bool transmitted, scalar_type VdotH, scalar_type LdotH, scalar_type VdotHLdotH, scalar_type orientedEta)
+    {
+        scalar_type FNG = fresnel_ndf * G1_over_2NdotV;
+        scalar_type factor = scalar_type(0.5);
+        if (transmitted)
+        {
+            const scalar_type VdotH_etaLdotH = (VdotH + orientedEta * LdotH);
+            // VdotHLdotH is negative under transmission, so this factor is negative
+            factor *= -scalar_type(2.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
+        }
+        return FNG * factor;
+    }
+
+    scalar_type DG1(scalar_type ndf, scalar_type G1_over_2NdotV)
+    {
+        return ndf * scalar_type(0.5) * G1_over_2NdotV;
+    }
```

Comment:
a reflective specialized overload shouldn't have its own code to implement, it should call the general function with constants and dummies
```cpp
const scalar_type dummy = 0.f;
return DG1(ndf,G1_over2NdotV,false,dummy,dummy,dummy,dummy,dummy);
```